File: js/core/HanGeulBunHae.js
import { HANGEUL_CHOSEONG, HANGEUL_JUNGSEONG, HANGEUL_JONGSEONG, SSANGJAEUM, GYEOPBATCHIM } from "../constant/hangeul/hangeul-mukja.js";

export const hanGeulBunHae = (char) => {
    const base = char.charCodeAt(0) - 0xAC00;
    if (base < 0 || base >= 0x2BA3) return null;

    const cho = Math.floor(base / (21 * 28));
    const jung = Math.floor((base % (21 * 28)) / 28);
    const jong = base % 28;

    const choseong = HANGEUL_CHOSEONG[cho];
    const jungseong = HANGEUL_JUNGSEONG[jung];
    const jongseong = HANGEUL_JONGSEONG[jong];

    return decompose({choseong, jungseong, jongseong});
}

function decompose({choseong, jungseong, jongseong}) {
    let doensori, gyeopbatchim;
    
    let cho = SSANGJAEUM.get(choseong) ?? choseong;
    if (SSANGJAEUM.get(choseong)) doensori = true;
    
    let jong = GYEOPBATCHIM.get(jongseong) ?? jongseong;
    if (GYEOPBATCHIM.get(jongseong)) {
        jongseong = jong.at(0);
        gyeopbatchim = jong.at(-1);
    }

    return {
            doensori,
            choseong: cho,
            jungseong,
            jongseong: jongseong,
            gyeopbatchim
    }
}

File: java/demo/src/main/java/com/junyeok/service/HanGeulBunHae.java
package com.junyeok.service;

import com.junyeok.constants.GyeopBatChim;
import com.junyeok.constants.HanGeulMukJa;
import com.junyeok.constants.SsangJaEum;
import com.junyeok.model.Geul;
import com.junyeok.model.HanGeul;

public class HanGeulBunHae {
    public static HanGeul bunhae(String c) {
        int base = c.charAt(0) - 0xAC00;
        if (base < 0 || base >= 0x2BA3) return null;

        int cho = base / (21 * 28);
        int jung = (base % (21 * 28)) / 28;
        int jong = base % 28;

        String choseong = HanGeulMukJa.HANGEUL_CHOSEONG[cho];
        String jungseong = HanGeulMukJa.HANGEUL_JUNGSEONG[jung];
        String jongseong = HanGeulMukJa.HANGEUL_JONGSEONG[jong];

        return HanGeul.builder()
            .choseong(choseong)
            .jungseong(jungseong)
            .jongseong(jongseong)
            .build();
    }
    public static Geul splitGyeop(HanGeul hangeul) {
        String doensori = "";
        String choseong = hangeul.getChoseong();
        String jongseong = hangeul.getJongseong();
        String jungseong = hangeul.getJungseong();
        String gyeop = "";

        SsangJaEum jaeum = SsangJaEum.fromChar(choseong);
        if (jaeum != null) {
            doensori = jaeum.getJaeum();
            choseong = jaeum.getJaeum();
        }
        GyeopBatChim batchim = GyeopBatChim.fromChar(jongseong);
        if (batchim != null) {
            jongseong = batchim.getFirst();
            gyeop = batchim.getSecond();
        }

        return HanGeul.builder()
            .doensori(doensori)
            .choseong(choseong)
            .jungseong(jungseong)
            .jongseong(jongseong)
            .gyeopbatchim(gyeop)
            .build();
    }
}